### ISSUE TITLE
[SUGGESTION] Possible Solution for relinking renamed/moved files

### ISSUE THREAD

**Issue by lloyd094** [CyanVoxel/TagStudio#36]
First: I don't have a full grasp as how two tags are implemented, so this may not be an option.

Can this issue be solved by creating a checksum based off the file and/or providing the file a hidden tag / database ID number, which would then continue with the file? The idea would be this number is unique, and can then be found later on if the file was previously in Tag Studio but later removed.

For example, if I took file `cat.png`, after it was already in Tag Studio and it was given `ID: 1`, then I removed it from my computer and later added that same file that had the hidden `ID: 1`, TagStudio would then know what that file is (or it's other associated tags). 

This implementation may require a local database or log file, which may be against your implementation idea, but it could help with recovery on that specific computer.

**Comment by CyanVoxel** [comment]
Are you referring to keeping some sort of unique ID inside of the file, via something like EXIF metadata?

Right now, TagStudio assigns each file with it's folder path a unique ID that it keeps track of inside an internal database. If the file is moved, and that path+filename key becomes orphaned - however TagStudio can search and relink that file if it finds a file with the same name in your library directory. There's lots of cases where this wouldn't work though, but we've got plans to implement additional methods of detecting when files are moved, modified, and renamed - including checksums and hashes.

Anyway, if you *were* referring to storing a unique id inside of files to help aide with this process, it's on our radar but is lower down the list of things to try, since it involves modifying users' files and will only work with files that support having that kind of metadata written to them. Either way, thank you for the suggestion!

**Comment by lloyd094** [comment]
Yes! The unique ID attached to the file itself if possible, and if not (or for files that don't support it) something with checksums to give the file a better chance to be recognized. Thank you!

**Comment by Trevo525** [comment]
Instead of modifying the files in any way, why not create a checksum of each file and use that as an id? When a file gets added that a checksum already exists, it could then be checked if it's a duplicate. If the file doesn't exist in it's original location anymore, then you can change the path, no harm no foul. If the file exists in both locations, then we can prompt the user for which file to keep (and delete the other), or maybe allow it keep both and one of them becomes the default for when you copy/paste. 

Just brainstorming. I wouldn't want TagStudio to edit the files in any way personally.

**Comment by Icosahunter** [comment]
I created an algorithm for relinking files for you here: https://github.com/Icosahunter/relink_utils_for_tagstudio
It can optionally take hashes for when you add that (it includes a simple function to create hashes and assumes you use this).
I made the interface independent from Tag Studio's implementation, it just takes file paths (and optionally a hash) as inputs.

**Comment by shfeat** [comment]
How about file's md5? 
Considering performance, we perform md5 based on the first 64k bytes

**Comment by yedpodtrzitko** [comment]
fyi there's [hashlib.file_digest](https://docs.python.org/3/library/hashlib.html#hashlib.file_digest) for calculating file hashes

**Comment by yedpodtrzitko** [comment]
one more thing to this topic - using [xxhash](https://xxhash.com) could be more performant than sha1 etc., no matter which way sha1 goes through the file

**Comment by shfeat** [comment]
> one more thing to this topic - using [xxhash](https://xxhash.com) could be more performant than sha1 etc., no matter which way sha1 goes through the file

wow, seem to be a better solution

**Comment by KillyMXI** [comment]
Preserving relevant discussion from Discord, with minor edits.
Happened at August 30th.

Start of the discussion: https://discord.com/channels/1229183630228848661/1229309667528806420/1278835404715458590

Computerdores — Today at 1:55 AM

> wouldn't it be possible to "just" store a hash of the file?
> and use that to detect moves / renames? 

CyanVoxel — Today at 1:56 AM

> Storing a hash of the file is one of the ideas included in the video and linked issues. It could be used to help relink a move, or detect a change, but not both at once. It's a tool we're planning on leveraging though

Killy — Today at 2:01 AM

> Similar thing that Git tries to solve a bit further.
> When you have a changeset, after detecting clear moves, if there are still deleted and added files, they can be compared for "strong similarity"
> But Git knows what file was like before
> 
> Without versioning, guessing the move + minor edit might be achieved by comparing known metadata or employing special hash functions (perceptual hash, media fingerprinting) (EDIT: https://en.wikipedia.org/wiki/Perceptual_hashing)

radu.m91 — Today at 2:27 AM

> Does the filesystem inode number change when the file is moved or edited? I suspect it doesn't. This means that while displaying the files, the program could check the filepath, modified time, content hash and inode number in order to detect if the file has been moved or edited, before displaying it. If it has been edited the inode number would be the same, but the content hash would differ. If it has been moved, the content hash of that inode would be the same, and the program could search for filepaths pointing to that inode in order to find the new location. (EDIT: https://en.wikipedia.org/wiki/Inode)

AiSO'); DROP TABLE users;-- — Today at 2:43 AM

> If a file on linux is moved into a different folder, renamed or its content changes, the inode remains the same. 

Killy — Today at 2:52 AM (reply to Killy)

> Side effect of having hashes of all files - easy to find exact duplicates. And similar media in case of perceptual hashes afaik, possibly even with adjustable level of similarity. That's how duplicate search works in Hydrus and https://github.com/ermig1979/AntiDupl

Killy — Today at 3:03 AM

> https://xxhash.com/

yedpodtrzitko — Today at 5:47 AM

> it would be good to concentrate the discussion and suggestions in the Issue which Cyan mentioned ( https://github.com/TagStudioDev/TagStudio/issues/36 ), nobody gonna find it here within a week among the stream of messages.

K — Today at 10:09 AM (reply to Killy)

> problem with hash it it long and kill ur hdd speed and cpu

Killy — Today at 12:16 PM (reply to K)

> You have to be smart about it.
> Avoid recomputing hashes.
> Use fast hashing algorithms. (That's why I linked xxhash.)
> It may still matter for big files though.
> The tools I've mentioned are very efficient on image files.
> 
> How things work on big files - can be seen in another media library application called "stash" - it is intended mostly for videos.
> It still computes file hash for all added files, among other things that take more time anyway. Only perceptual hash is an option that is not enabled by default. (EDIT: https://github.com/stashapp/stash/)

K — Today at 1:03 PM (reply to Killy)

> > xxhash
> 
> First I hear of this, does people use or support this hash though?
> I just had a deeper look at their website, I recognize some software there
> hmm CLI only software from look of their git release

Killy — Today at 1:29 PM (reply to K)

> It's an algorithm with implementations or library bindings in different languages.
> Intended for integration into your software - TagStudio for example.
> Its entire concern is to return a hash for a given file as fast as possible.
> The usage of that hash is a concern of the client application - be that move detection, duplicate detection, integrity check or whatever else... 

Killy — Today at 1:38 PM

> There are a lot more non-cryptographic hash functions (read: not slowed down artificially)
> Many (not all) are mentioned at https://en.wikipedia.org/wiki/List_of_hash_functions
> But xxhash seems best known, well regarded

**Comment by felindev** [comment]
I looked a bit into implementing something like this. I've noticed that when restarting app it tries to refresh library, possibly adding duplicate of existing image which then makes it harder to replace it (UNIQUE on path, new entry). I'm thinking best course of action would be to implement hashing on refresh/adding to database (with modified time as indicator that hashes need to be updated), checking beforehand if given hash already exists. 
This way we get de-duplication automatically - which might not be intended effect. I'm thinking, would it be a good idea to add pop-up for user to ask which file location to use removing other ones. Optionally adding some kind of discriminator to let user keep both copies though it might not be worth the effort - TagStudio is media management. Keeping exact copies in library doesn't seem useful. 
I have some experience with perceptual hashes, though like in discussion linked above, it'd be a good idea to defer it to later deduplication. It will quickly get out of hand on larger datasets.


What do you think? Should we let user keep exact duplicates or force them to keep only one copy. If we allow copies, do we keep it as one entry or multiple ones with separate tags?

**Comment by Trevo525** [comment]
> What do you think? Should we let user keep exact duplicates or force them to keep only one copy. If we allow copies, do we keep it as one entry or multiple ones with separate tags?

To me, the whole point of using TagStudio would be to use it in place of File Explorer for certain files. So, deduplication is the only way that makes sense. Because who cares where files are stored if you are searched based on tags.

That's just my two cents.

**Comment by CyanVoxel** [comment]
> I looked a bit into implementing something like this. I've noticed that when restarting app it tries to refresh library, possibly adding duplicate of existing image which then makes it harder to replace it (UNIQUE on path, new entry). I'm thinking best course of action would be to implement hashing on refresh/adding to database (with modified time as indicator that hashes need to be updated), checking beforehand if given hash already exists. This way we get de-duplication automatically - which might not be intended effect. I'm thinking, would it be a good idea to add pop-up for user to ask which file location to use removing other ones. Optionally adding some kind of discriminator to let user keep both copies though it might not be worth the effort - TagStudio is media management. Keeping exact copies in library doesn't seem useful. I have some experience with perceptual hashes, though like in discussion linked above, it'd be a good idea to defer it to later deduplication. It will quickly get out of hand on larger datasets.
> 
> What do you think? Should we let user keep exact duplicates or force them to keep only one copy. If we allow copies, do we keep it as one entry or multiple ones with separate tags?

The automatic refresh on application start is currently an experimental behavior of the `main`/incomplete 9.5 branch, which is also missing the de-duplication fixes introduced in the 9.4 branch. This results in new duplicate entry issues that are not present on 9.4, which is something that needs to be addressed by us at some point.

As for allowing "duplicate" files inside separate locations, this is 100% what we want for TagStudio and is the intended behavior to keep. This allows TagStudio to be used on top of existing project file structures. One example would be having a root "Projects" folder which includes two subfolders, "Project 1" and Project 2". Within each one of those subfolders, you have a blank "README.md" file that may have the same file creation and modification date metadata. If we restricted file contents and/or metadata to be unique for identification, then we would be left with the incorrect assumption that these are the "same" file and should share the same tags, even if they're files for otherwise completely different projects.
TLDR: while hashes will be necessary, file paths are also required to allow for otherwise "duplicate" files.

>  I'm thinking, would it be a good idea to add pop-up for user to ask which file location to use removing other ones. Optionally adding some kind of discriminator to let user keep both copies though it might not be worth the effort - TagStudio is media management.

Also in case I'm reading into this incorrectly, we do *not* wish to support multiple entries for the same file *in the same location*. At least, at the per-library level. This is what having multiple libraries would allow for, though.

**Comment by mm12** [comment]
> How about file's md5? Considering performance, we perform md5 based on the first 64k bytes

This is really the correct way

**Comment by GaiusScotius** [comment]
I've just started looking at Tag Studio and was surprised that the SQLite database seems to identify files by file name and can thus lose track of them if they are moved.

Another method of providing a unique file identifier, at least on POSIX compliant systems, is to use the combination of disk identifier (on Linux systems this is a UUID, on others you may have to use the /dev entry or mount point); the file's inode; and (because inodes may be reused after a file is deleted) the the file creation date. This combination is unique to the file and does not change if the file is moved. Another advantage is that hard links (i.e those made by the ln command, not ln -s) share the same inode.

Using <DISK ID>,<INODE>,<CREATION DATE> as a unique ID in the "entries" table could go a long way to solving the issue of moving files. You can't open file by inode, that is not allowed as it would bypass access control, but if an attempt to open a file by its name fails other than for reasons of permissions, running "find -inum" will find the file if it still exists. You do, of course, have to have some idea of which directory find should search (the user's home directory is a good bet, particularly if you let find following soft links and mounts) and find mightn't be particularly quick in very large directory trees, but you could always ask the user if Tag Studio should try looking for the file.

**Comment by Computerdores** [comment]
Something else I have stumbled across is extended attributes (or xattrs) which are available on all major linux/mac filesystems[^2] and alternate data streams for NTFS - these basically allow attaching extra metadata to a file without modifying it's contents[^1]. We could use this to attach a unique identifier to a file, which would hopefully mean that even when a file is renamed and changed, Tagstudio would still be able to correctly identify it.

I haven't tested this yet, so I don't know how well this is handled when moving files between NTFS and the linux filesystems, but even if these attributes get lost entirely, then TS could just regenerate them (this is assuming no one would rename+move+move between filesystems at the same time).

Assuming this works as I think it does, I would propose using this as the main method as it would be the most reliable and fast one, and then falling back to file hashes if these attributes are not present anymore.

[^1]: Not modifying file contents (unless there is a HUGE warning; see file delete button) has been a long standing rule for any features we introduce.
[^2]: See [Arch Wiki](https://wiki.archlinux.org/title/Extended_attributes#File_systems)

**Comment by GaiusScotius** [comment]
I’m not convinced adding a unique ID as metadata stored in an xattrs() value is the way to go. The reason is that they are not searchable; once you have a file handle you can get value, but you need to get the file first. With the inode (plus volume ID) you can open the file irrespective of whether it’s file name is the same.

**Comment by Computerdores** [comment]
> I’m not convinced adding a unique ID as metadata stored in an xattrs() value is the way to go. The reason is that they are not searchable

Could you elaborate further? The use case here is that when during the routine scan of the library, files are found that TS does not know about and some files it does know about are not found anymore it can relink any moved files as reliably as possible. The current ideas all involve some kind of content comparison of the file which would not only also require a file handle, but most likely require reading the entire file to compute a hash.

Note: Currently this only happens when a path does not match, so it would not catch swapping the filenames of two files, this approach would also make it possible to detect that with relatively low cost.

**Comment by GaiusScotius** [comment]
The problem with storing anything as metadata attached to a file is that there’s no mechanism in the filesystem itself to find/open files based on metadata; that includes the file’s inode.

If during a scan you find a file is missing, the question is whether it’s been moved or deleted. That is relatively easy to check if you know the file’s inode; the command “find <dir> -inum <inode>” will return all directory entries that have the given inode. If nothing is returned, the file has been deleted.

If something is returned the file has almost certainly been moved. Almost certainly because inodes can be recycled after a file is deleted. To determine if the inode is now being used by a different file it needs to be
combined with some other value. Most modern files systems store a “birth” or “creation date”, which achieves this. Calling stat() on the file will return it. 

The two steps can be combined using the exec parameter of find.

Equally one could code the equivalent of “find” and have it look for a TS specific metadata attribute.

This a Unix/Posix approach. I don’t know if Windows has the same capabilities.

**Comment by Computerdores** [comment]
The problems I see with this approach are:
- as you said "This a Unix/Posix approach". TagStudio needs to support windows and there may even be unix filesystems that don't use inodes (not sure on this, but I know that xattrs are available on all major unix filesystems and an equivalent is also available on NTFS). But mainly: Reengineering a different solution for windows would be significant extra work (assuming something like this is even possible on NTFS)
- as you also said, this approach would in every case need a way to make sure that the candidate for having been moved is actually the same file, where as the xattrs approach would only require this if the xattr is not present anymore
- moving a TS library to another instance of the same FS would need to be detected (all inodes would be different, every file would be seen as moved, all inodes need to be reacquired for all entries). Meanwhile xattrs would simply be preserved and no further action would be necessary
- If the inode is not found in any file anymore, the inode could also have changed (I am not aware of any guarantee that a file of same contents will always have the same inode, so you could always construct cases where this will happen) this means that you can't assume the file has been deleted and have to check all files anyway
- I see no advantage to being able to search for the inode, since all files in the library have to be checked anyway on a rescan (because `example1.jpg` may have been moved to `example2.jpg` replacing the original `example2.jpg`; which currently would be detected as deleting `example1.jpg`, where as really `example2.jpg` was deleted and `example1.jpg` was moved)

Also more generally: Afaik no guarantees related to what you propose doing with inodes exist for them (which makes me weary of using them this way, because I don't want to be [longtimeuser4](https://xkcd.com/1172/)), where as we would be using xattrs for the exact purpose they are intended for

**Comment by zitongcharliedeng** [comment]
Saw some UUID/xattr/inode mentions in this thread and thought I'd chip in with a different angle.

Has anyone considered making the library folder a Git repository under the hood — or leveraging an existing one? Many users already git-track their vaults/libraries/project folders. TagStudio could detect an existing `.git/` and use `git diff -M` for rename/move detection, or offer to `git init` if there isn't one. Git infers renames by comparing tree snapshots with similarity detection — no identity stored in files, no xattrs, no inodes, works on every OS and filesystem.

Beyond relinking, this would also give you:
- Tag/metadata change history (commit the DB or a text export after each change)
- Undo for metadata operations
- Library sync between machines via push/pull

The main limitation today is modified binaries — Git can't detect a re-encoded video with a new filename. But Git is one of the most actively developed tools in existence, and any future improvements to its binary handling (like better binary diffing, perceptual matching, or LFS-level tracking) would flow directly to TagStudio for free. Coupling to Git means gaining parity with an enormous ecosystem rather than maintaining a custom solution.

**Comment by mm12** [comment]
> Saw some UUID/xattr/inode mentions in this thread and thought I'd chip in with a different angle.
> 
> Has anyone considered making the library folder a Git repository under the hood — or leveraging an existing one? Many users already git-track their vaults/libraries/project folders. TagStudio could detect an existing `.git/` and use `git diff -M` for rename/move detection, or offer to `git init` if there isn't one. Git infers renames by comparing tree snapshots with similarity detection — no identity stored in files, no xattrs, no inodes, works on every OS and filesystem.
> 
> Beyond relinking, this would also give you:
> 
>     * Tag/metadata change history (commit the DB or a text export after each change)
> 
>     * Undo for metadata operations
> 
>     * Library sync between machines via push/pull
> 
> 
> The main limitation today is modified binaries — Git can't detect a re-encoded video with a new filename. But Git is one of the most actively developed tools in existence, and any future improvements to its binary handling (like better binary diffing, perceptual matching, or LFS-level tracking) would flow directly to TagStudio for free. Coupling to Git means gaining parity with an enormous ecosystem rather than maintaining a custom solution.

I considered it... but probably not a very good idea.
